src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,529 @@
+
+        deleted_servers = old_servers - new_servers
+        if deleted_servers:
+            logger.info(f"servers deleted: {deleted_servers}")
+
+        self.send_request_to_update_servers(self.servers)
+
+    def send_request_to_update_servers(self, servers: List[str]):
+        AUTHORIZATION_TYPE = "Basic"
+        USERNAME = "lightning"
+        AUTO_SCALER_AUTH_PASSWORD = os.environ.get("AUTO_SCALER_AUTH_PASSWORD", "")
+
+        try:
+            param = f"{USERNAME}:{AUTO_SCALER_AUTH_PASSWORD}".encode()
+            data = b64encode(param).decode("utf-8")
+        except (ValueError, UnicodeDecodeError) as e:
+            raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                detail="Invalid authentication credentials",
+                headers={"WWW-Authenticate": "Basic"},
+            ) from e
+        headers = {
+            "accept": "application/json",
+            "username": USERNAME,
+            "Authorization": AUTHORIZATION_TYPE + " " + data,
+        }
+        response = requests.put(f"{self.url}/system/update-servers", json=servers, headers=headers, timeout=10)
+        response.raise_for_status()
+
+
+class AutoScaler(LightningFlow):
+    """A LightningFlow component that handles all the servers and uses load balancer to spawn up and shutdown based
+    on current requests in the queue.
+
+    Args:
+        min_replicas: Number of works to start when app initializes.
+        max_replicas: Max numbers of works to spawn to handle the incoming requests.
+        autoscale_interval: Number of seconds to wait before checking whether to upscale or downscale the works.
+        max_batch_size: Number of requests to process at once.
+        timeout_batch: Number of seconds to wait before sending the requests to process.
+        downscale_threshold: Lower limit to determine when to stop works.
+        upscale_threshold: Upper limit to determine when to spawn up a new work.
+        worker_url: Default=api/predict. Provide the REST API path
+        input_schema:
+        output_schema:
+
+        .. doctest::
+
+            >>> from lightning_app.components import AutoScaler
+            >>> from lightning_app import LightningApp
+            >>> app = LightningApp(
+            ...     AutoScaler(
+            ...         MyPythonServer,  # noqa: F821
+            ...         min_replicas=1,
+            ...         max_replicas=4,
+            ...         autoscale_interval=10,
+            ...     )
+            ... )
+    """
+
+    def __init__(
+        self,
+        work_cls: Optional[type] = None,
+        min_replicas: int = 1,
+        max_replicas: int = 4,
+        autoscale_interval: int = 1 * 10,
+        max_batch_size: int = 8,
+        timeout_batch: float = 2,
+        downscale_threshold: Optional[int] = None,
+        upscale_threshold: Optional[int] = None,
+        worker_url: str = None,
+        input_schema: Any = Dict,
+        output_schema: Any = Dict,
```

Comment:
would be `{in|out}put_type` cleaner for the meaning? or `{in|out}put_schema_type`

Reply:
I'll have them as `{in,out}put_type` to make them consistent across other components